pymc/tests/test_gp.py

```diff
@@ -1204,29 +1237,54 @@ def test_plot_gp_dist_warn_nan(self):
         pass
 
 
-class TestCircular:
-    def test_1d_tau1(self):
-        X = np.linspace(0, 1, 10)[:, None]
-        etalon = 0.600881
-        with pm.Model():
-            cov = pm.gp.cov.Circular(1, 1, tau=5)
-        K = aesara.function([], cov(X))()
-        npt.assert_allclose(K[0, 1], etalon, atol=1e-3)
-        K = aesara.function([], cov(X, X))()
-        npt.assert_allclose(K[0, 1], etalon, atol=1e-3)
-        # check diagonal
-        Kd = aesara.function([], cov(X, diag=True))()
-        npt.assert_allclose(np.diag(K), Kd, atol=1e-5)
+class TestKmeansInducing:
+    def setup_method(self):
+        self.centers = (-5, 5)
+        self.x = np.concatenate(
+            (self.centers[0] + np.random.randn(500), self.centers[1] + np.random.randn(500))
+        )
 
-    def test_1d_tau2(self):
-        X = np.linspace(0, 1, 10)[:, None]
-        etalon = 0.691239
-        with pm.Model():
-            cov = pm.gp.cov.Circular(1, 1, tau=4)
-        K = aesara.function([], cov(X))()
-        npt.assert_allclose(K[0, 1], etalon, atol=1e-3)
-        K = aesara.function([], cov(X, X))()
-        npt.assert_allclose(K[0, 1], etalon, atol=1e-3)
-        # check diagonal
-        Kd = aesara.function([], cov(X, diag=True))()
-        npt.assert_allclose(np.diag(K), Kd, atol=1e-5)
+    def test_kmeans(self):
+        X = self.x[:, None]
+        Xu = pm.gp.util.kmeans_inducing_points(2, X).flatten()
+        npt.assert_allclose(np.asarray(self.centers), np.sort(Xu), atol=0.1)
+
+        X = at.as_tensor_variable(self.x[:, None])
+        Xu = pm.gp.util.kmeans_inducing_points(2, X).flatten()
+        npt.assert_allclose(np.asarray(self.centers), np.sort(Xu), atol=0.1)
+
+    def test_kmeans_raises(self):
+        with pytest.raises(TypeError):
+            Xu = pm.gp.util.kmeans_inducing_points(2, "str is the wrong type").flatten()
+
+
+class TestReplaceWithValues:
+    def test_basic_replace(self):
+        with pm.Model() as model:
+            a = pm.Normal("a")
+            b = pm.Normal("b", mu=a)
+            c = a * b
+
+        (c_val,) = pm.gp.util.replace_with_values(
+            [c], replacements={"a": 2, "b": 3, "x": 100}, model=model
```

Comment:
Any reason to base this function on the variables names as opposed to the variables themselves?

Reply:
No particular reason, it just struck me as the most natural, but I'm not extremely familiar with Aesara and its usage patterns.  I suppose it'd be just as well to use the variables themselves as keys too.  Is there a reason that would be preferred?  Happy to switch to that though.

Reply:
In general I think we are trying to move away from our dependency on variable names but this is probably never going to happen so... I guess either way.

About duplicating code, if this is a once per sampling kind of thing, I think it is fine.

More importantly tests are passing and we should try to merge soon as it is quite a large PR. We can always come back later for final polishing

Reply:
If we're trying to move away from variable string names I'll go ahead and change it.  Out of curiosity, what's the benefit for doing so?

Also, for `gp.predict`, it's not even once per sample, it's just once.  You'd make your GP model, take the MAP estimate say, and then do something like:

```python
with model:
    mu, cov = gp.predict(X, point=map_estimate)

# plot prediced mean
plt.plot(mu)
# +- 2 sigma uncertainty
plt.plot(mu + 2*np.sqrt(np.diag(cov)), 'b')
plt.plot(mu - 2*np.sqrt(np.diag(cov)), 'b')
```

Should `replace_with_values` live in `gp.util`?  Or should it move to somewhere general, like `aesaraf`?  There isn't anything necessarily GP specific about it -- it could be of general use. 

Sounds good, and yes sorry that it is a large PR!  I'm happy to talk through anything too.

Reply:
> If we're trying to move away from variable string names I'll go ahead and change it.  Out of curiosity, what's the benefit for doing so?

It makes somethings easier like graph manipulations or variable inspection (e.g, you don't need access to the model object to know what you are dealing with). Also we don't need to worry about the transformed names which are another nuisance in a lot of the codebase. 

Less important perhaps there is no real limitation at the Aesara level that variables must have unique names.

> Also, for `gp.predict`, it's not even once per sample, it's just once.  You'd make your GP model, take the MAP estimate say, and then do something like:
> 
> ```python
> with model:
>     mu, cov = gp.predict(X, point=map_estimate)
> 
> # plot prediced mean
> plt.plot(mu)
> # +- 2 sigma uncertainty
> plt.plot(mu + 2*np.sqrt(np.diag(cov)), 'b')
> plt.plot(mu - 2*np.sqrt(np.diag(cov)), 'b')
> ```
> 
> Should `replace_with_values` live in `gp.util`?  Or should it move to somewhere general, like `aesaraf`?  There isn't anything necessarily GP specific about it -- it could be of general use. 

Is there a case were you might want to evaluate for more than a single point? Is it conceptually similar to posterior predictive?

If so you may want to check if any of that code in sample_posterior_predictive may make more sense here.

Reply:
> It makes somethings easier like graph manipulations or variable inspection (e.g, you don't need access to the model object to know what you are dealing with). Also we don't need to worry about the transformed names which are another nuisance in a lot of the codebase.

Gotcha, that makes a lot of sense, especially the transformed/untransformed distinction -- ideally that should stay under the hood. 

> Is there a case were you might want to evaluate for more than a single point? Is it conceptually similar to posterior predictive sampling?

Never say never, but I doubt it.  `gp.predict` is really just a convenience method for getting the conditional mean and variance given the MAP estimate.  I actually did start with `sample_posterior_predictive` to figure out how to make this function, since that's where I'd stolen `draw_values` from originally!  But it looked like draw_values had been refactored into component pieces.